# Déjame escuchar música

Los datos están almacenados en el archivo `/datasets/music_project_en.csv`.

¡Ahora sí, manos al código!


## Etapa 1. Descripción de los datos <a id='data_review'></a>

Abre los datos y examínalos.

Etapa 1.1. Necesitarás `pandas`, así que impórtalo.

In [2]:
import pandas as pd #Importar pandas


Etapa 1.2. Lee el archivo `music_project_en.csv` de la carpeta `/datasets/` y guárdalo en la variable `df`:

In [3]:
df = pd.read_csv("/datasets/music_project_en.csv")# Lee el archivo y almacénalo en df


Etapa 1.3. Muestra las 10 primeras filas de la tabla:

In [4]:
df.head(10)# Obtén las 10 primeras filas de la tabla df


,userID,Track,artist,genre,City,time,Day
0,FFB692EC,Kamigata To Boots,The Mass Missile,rock,Shelbyville,20:28:33,Wednesday
1,55204538,Delayed Because of Accident,Andreas Rönnberg,rock,Springfield,14:07:09,Friday
2,20EC38,Funiculì funiculà,Mario Lanza,pop,Shelbyville,20:58:07,Wednesday
3,A3DD03C9,Dragons in the Sunset,Fire + Ice,folk,Shelbyville,08:37:09,Monday
4,E2DC1FAE,Soul People,Space Echo,dance,Springfield,08:34:34,Monday
5,842029A1,Chains,Obladaet,rusrap,Shelbyville,13:09:41,Friday
6,4CB90AA5,True,Roman Messer,dance,Springfield,13:00:07,Wednesday
7,F03E1C1F,Feeling This Way,Polina Griffith,dance,Springfield,20:47:49,Wednesday
8,8FA1D3BE,L’estate,Julia Dalia,ruspop,Springfield,09:17:40,Friday
9,E772D5C0,Pessimist,NaN,dance,Shelbyville,21:20:49,Wednesday


Etapa 1.4. Obtén la información general sobre la tabla con el método info().

In [5]:
df.info()# Obtén la información general sobre nuestros datos


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65079 entries, 0 to 65078
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0     userID  65079 non-null  object
 1   Track     63736 non-null  object
 2   artist    57512 non-null  object
 3   genre     63881 non-null  object
 4     City    65079 non-null  object
 5   time      65079 non-null  object
 6   Day       65079 non-null  object
dtypes: object(7)
memory usage: 3.5+ MB


Estas son nuestras observaciones sobre la tabla. Contiene siete columnas que almacenan los mismos tipos de datos: `object`.

Según la documentación:
- `' userID'`: identificador del usuario;
- `'Track'`: título de la canción;
- `'artist'`: nombre del artista;
- `'genre'`: género de la canción;
- `'City'`: ciudad del usuario;
- `'time'`: la hora exacta en la que se reprodujo la canción;
- `'Day'`: día de la semana.

Podemos ver dos problemas con el estilo en los encabezados de la tabla:
1. Algunos encabezados están en mayúsculas, otros en minúsculas.
2. `Identifica tú mismo el segundo problema y escríbelo aquí.`
Hay espacios en los nombres de las columnas que pueden generar problemas, por ejemplo en " userID"



### Escribe algunas observaciones por tu parte. Contesta a las siguientes preguntas: <a id='data_review_conclusions'></a>

`1.   ¿Qué tipo de datos hay en las filas? ¿Cómo podemos saber qué almacenan las columnas?`

`2.   ¿Hay suficientes datos para proporcionar respuestas a nuestra hipótesis o necesitamos más información?`

`3.   ¿Notaste algún problema en los datos, como valores ausentes, duplicados o tipos de datos incorrectos?`

Escribe aquí tus respuestas:

1. Los datos son de tipo texto (object) y podemos notar que cada fila representa una reproduccion de cancion, podemos saber lo que las columnas almacenan por el df.head() que nos muestra las primeras 10 filas de la tabla

2. Con mas de 65 mil entradas hay suficiente informacion para proporcionar respuestas a nuestras hipotesis

3. si, hay valores NaN y podemos notar en df.info que las columnas track, artist y genre tienen problemas con los datos 

## Etapa 2. Preprocesamiento de los datos <a id='data_preprocessing'></a>

Tu objetivo aquí es preparar los datos para analizarlos.
El primer paso es resolver los problemas con los encabezados. Después podemos avanzar a los valores ausentes y duplicados. ¡Empecemos!

Vamos a corregir el formato en los encabezados de la tabla.


### Estilo del encabezado <a id='header_style'></a>
Etapa 2.1. Muestra los encabezados de la tabla (los nombres de las columnas):

In [6]:
df.columns# Muestra los nombres de las columnas


Index(['  userID', 'Track', 'artist', 'genre', '  City  ', 'time', 'Day'], dtype='object')

Vamos cambiar los encabezados de la tabla siguiendo las reglas estilísticas convencionales:
*   Todos los caracteres deben ser minúsculas.
*   Elimina los espacios.
*   Si el nombre tiene varias palabras, utiliza snake_case, es decir, añade un guion bajo ( _ ) entre las palabras en lugar de un espacio.



Etapa 2.2. Utiliza el bucle for para iterar sobre los nombres de las columnas y poner todos los caracteres en minúsculas. Cuando hayas terminado, vuelve a mostrar los encabezados de la tabla:

In [7]:
new_columns = []
for cols in df.columns:
    new_columns.append(cols.lower())

df.columns = new_columns

df.columns# Bucle que itera sobre los encabezados y los pone todos en minúsculas


Index(['  userid', 'track', 'artist', 'genre', '  city  ', 'time', 'day'], dtype='object')

Etapa 2.3. Ahora, utilizando el mismo método, elimina los espacios al principio y al final de los nombres de las columnas y muestra los nombres de las columnas de nuevo:

In [8]:
new_columns = []
for cols in df.columns:
    new_columns.append(cols.strip())

df.columns = new_columns

df.columns# Bucle que itera sobre los encabezados y elimina los espacios


Index(['userid', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')

Etapa 2.4. Necesitamos aplicar la regla de snake_case en la columna `userid`. Debe ser `user_id`. Cambia el nombre de esta columna y muestra los nombres de todas las columnas cuando hayas terminado.

In [9]:
df.rename(columns = {"userid": "user_id"}, inplace = True)

df.columns# Cambia el nombre de la columna "userid"


Index(['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')

Etapa 2.5. Comprueba el resultado. Muestra los encabezados una vez más:

In [10]:
df.columns # Comprueba el resultado: lista de encabezados


Index(['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')

### Valores ausentes <a id='missing_values'></a>
 Etapa 2.5. Primero, encuentra el número de valores ausentes en la tabla. Debes utilizar dos métodos para obtener el número de valores ausentes.

In [19]:
df.isna().sum()

df.info()# Calcula el número de valores ausentes

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65079 entries, 0 to 65078
Data columns (total 7 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   user_id  65079 non-null  object
 1   track    65079 non-null  object
 2   artist   65079 non-null  object
 3   genre    65079 non-null  object
 4   city     65079 non-null  object
 5   time     65079 non-null  object
 6   day      65079 non-null  object
dtypes: object(7)
memory usage: 3.5+ MB


Etapa 2.6. Sustituye los valores ausentes en las columnas `'track'`, `'artist'` y `'genre'` con el string `'unknown'`.

1. Crea una lista llamada columns_to_replace que contenga los nombres de las columnas 'track', 'artist' y 'genre'.

2. Usa un bucle for para iterar sobre cada columna en columns_to_replace.

3. Dentro del bucle, sustituye los valores ausentes en cada columna con el string `'unknown'`.


In [12]:
columns_to_replace = ["track", "artist", "genre"]

for cols in columns_to_replace:
    df[cols] = df[cols].fillna("unknown")
 # Bucle en los encabezados reemplazando los valores ausentes con 'unknown'


Etapa 2.7. Ahora comprueba el resultado para asegurarte de que no falten valores ausentes por reemplazar en el conjunto de datos. Para ello, cuenta los valores ausentes una vez más.

In [13]:
df[columns_to_replace].isna().sum() # Cuenta los valores ausentes


track     0
artist    0
genre     0
dtype: int64

### Duplicados <a id='duplicates'></a>
Etapa 2.8. Encuentra el número de duplicados explícitos en la tabla. Una vez más, debes aplicar dos métodos para obtener la cantidad de duplicados explícitos.

In [20]:
df.duplicated().sum()

total = len(df) 
unique = len(df.drop_duplicates())
duplicates = total - unique
print(duplicates)# Cuenta los duplicados explícitos

3826


Etapa 2.9. Ahora, elimina todos los duplicados. Para ello, llama al método que hace exactamente esto.

In [21]:
df = df.drop_duplicates()# Elimina los duplicados explícitos

Etapa 2.10. Comprobemos ahora si conseguimos eliminar todos los duplicados. Cuenta los duplicados explícitos una vez más para asegurarte de haberlos eliminado todos:

In [22]:
df.duplicated().sum()# Comprueba de nuevo si hay duplicados

0

Ahora queremos deshacernos de los duplicados implícitos en la columna `genre`. Por ejemplo, el nombre de un género se puede escribir de varias formas. Dichos errores también pueden afectar al resultado.

Etapa 2.11. Primero debemos mostrar una lista de nombres de géneros únicos, por orden alfabético. Para ello:
1. Extrae la columna `genre` del DataFrame.
2. Llama al método que devolverá todos los valores únicos en la columna extraída.


In [23]:
sorted(df["genre"].unique())# Inspecciona los nombres de géneros únicos


['acid',
 'acoustic',
 'action',
 'adult',
 'africa',
 'afrikaans',
 'alternative',
 'ambient',
 'americana',
 'animated',
 'anime',
 'arabesk',
 'arabic',
 'arena',
 'argentinetango',
 'art',
 'audiobook',
 'avantgarde',
 'axé',
 'baile',
 'balkan',
 'beats',
 'bigroom',
 'black',
 'bluegrass',
 'blues',
 'bollywood',
 'bossa',
 'brazilian',
 'breakbeat',
 'breaks',
 'broadway',
 'cantautori',
 'cantopop',
 'canzone',
 'caribbean',
 'caucasian',
 'celtic',
 'chamber',
 'children',
 'chill',
 'chinese',
 'choral',
 'christian',
 'christmas',
 'classical',
 'classicmetal',
 'club',
 'colombian',
 'comedy',
 'conjazz',
 'contemporary',
 'country',
 'cuban',
 'dance',
 'dancehall',
 'dancepop',
 'dark',
 'death',
 'deep',
 'deutschrock',
 'deutschspr',
 'dirty',
 'disco',
 'dnb',
 'documentary',
 'downbeat',
 'downtempo',
 'drum',
 'dub',
 'dubstep',
 'eastern',
 'easy',
 'electronic',
 'electropop',
 'emo',
 'entehno',
 'epicmetal',
 'estrada',
 'ethnic',
 'eurofolk',
 'european',
 'expe

Etapa 2.12. Vamos a examinar la lista para identificar **duplicados implícitos** del género `hiphop`, es decir, nombres mal escritos o variantes que hacen referencia al mismo género musical.

Los duplicados que encontrarás son:

* `hip`  
* `hop`  
* `hip-hop`  

Para solucionarlo, vamos a crear una función llamada `replace_wrong_values()`.


1. Define una función llamada `replace_wrong_values()` que reciba los siguientes parámetros:

* `df`: el DataFrame a modificar
* `column`: el nombre de la columna a trabajar
* `wrong_values`: una lista con los valores incorrectos
* `correct_value`: el valor correcto para reemplazar

2. Dentro de la función, usa un bucle `for` para iterar sobre cada valor incorrecto y aplicar `.replace()`.


In [33]:
def replace_wrong_values(df, column, wrong_values, correct_value):
    for wrong in wrong_values:
        df[column] = df[column].replace(wrong, correct_value)# Función para reemplazar los duplicados implícitos


Etapa 2.13. Ahora, llama a la función pasando:

* `df` como el DataFrame
* `'genre'` como nombre de columna
* `['hip', 'hop', 'hip-hop']` como lista de valores incorrectos
* `'hiphop'` como valor correcto


In [35]:
replace_wrong_values(df, "genre", ["hip", "hop", "hip-hop"], "hiphop")# Elimina los duplicados implícitos


Etapa 2.14. Asegúrate de que los nombres duplicados se hayan eliminado. Muestra la lista de valores únicos de la columna `'genre'` una vez más:

In [36]:
df["genre"].unique()# Comprueba de nuevo los duplicados implícitos


array(['rock', 'pop', 'folk', 'dance', 'rusrap', 'ruspop', 'world',
       'electronic', 'unknown', 'alternative', 'children', 'rnb',
       'hiphop', 'jazz', 'postrock', 'latin', 'classical', 'metal',
       'reggae', 'triphop', 'blues', 'instrumental', 'rusrock', 'dnb',
       'türk', 'post', 'country', 'psychedelic', 'conjazz', 'indie',
       'posthardcore', 'local', 'avantgarde', 'punk', 'videogame',
       'techno', 'house', 'christmas', 'melodic', 'caucasian',
       'reggaeton', 'soundtrack', 'singer', 'ska', 'salsa', 'ambient',
       'film', 'western', 'rap', 'beats', "hard'n'heavy", 'progmetal',
       'minimal', 'tropical', 'contemporary', 'new', 'soul', 'holiday',
       'german', 'jpop', 'spiritual', 'urban', 'gospel', 'nujazz',
       'folkmetal', 'trance', 'miscellaneous', 'anime', 'hardcore',
       'progressive', 'korean', 'numetal', 'vocal', 'estrada', 'tango',
       'loungeelectronic', 'classicmetal', 'dubstep', 'club', 'deep',
       'southern', 'black', 'folkrock

## Etapa 3. Análisis

### Tarea: Comparar el comportamiento de los usuarios en las dos ciudades <a id='activity'></a>

Queremos analizar si hay diferencias en la cantidad de canciones reproducidas en Springfield y Shelbyville. Para ello, usaremos los datos de dos días de la semana: lunes y viernes.

Compararemos cuántas canciones se escucharon en cada ciudad durante esos días para identificar posibles patrones de comportamiento.

Sigue estos tres pasos para organizar tu análisis:

- Dividir: agrupa los datos por ciudad.

- Aplicar: cuenta cuántas canciones se reproducen en cada grupo.

- Combinar: presenta los resultados de forma que se puedan comparar fácilmente ambas ciudades.

Repite este proceso por separado para cada uno de los dos días.

Etapa 3.1.
Cuenta cuántas canciones se reprodujeron en cada ciudad utilizando la columna `'track'` como referencia.

In [37]:
city_track_count = df.groupby("city")["track"].count()
print(city_track_count)# Cuenta las canciones reproducidas en cada ciudad


city
Shelbyville    18512
Springfield    42741
Name: track, dtype: int64


Etapa 3.2. Redacta brevemente tus observaciones sobre los resultados.

¿Qué diferencias encontraste entre Springfield y Shelbyville? ¿A qué podrían deberse?

Escribe tus observaciones aquí.

Springfield escucha mas del doble de canciones que Shelbyville, podria ser por la cantidad de usuarios, o porque en general tienen mas actividad

Etapa 3.3.
Agrupa los datos por día de la semana y cuenta cuántas canciones se reprodujeron los lunes y viernes.



In [39]:
df_filtered = df[df["day"].isin(["Monday", "Friday"])]
song_by_day = df_filtered.groupby("day")["track"].count()
print(song_by_day)# Calcula las canciones reproducidas en cada uno de los dos días


day
Friday    21840
Monday    21354
Name: track, dtype: int64


Etapa 3.4. Describe brevemente qué observaste al comparar los lunes y viernes.

¿Hubo un día con más actividad? ¿Cambia algo si analizas cada ciudad por separado?

Escribe tus observaciones aquí.

El viernes hay una actividad musical ligeramente mayor que en lunes, el comportamiento musical entre ambas ciudades es similar independientemente del dia de la semana

Etapa 3.5

Ahora vamos a combinar dos criterios: día y ciudad.

Crea una función llamada `number_tracks()` que reciba dos parámetros:

* `day`: un día de la semana (por ejemplo, `'Monday'`)
* `city`: el nombre de una ciudad (por ejemplo, `'Springfield'`)

Dentro de la función:

1. Filtra el DataFrame por el día.
2. Luego, filtra por la ciudad.
3. Cuenta cuántas veces aparece `'user_id'` en ese filtro.
4. Devuelve ese número como resultado.


In [41]:
# Declara la función number_tracks() con dos parámetros: day= y city=.
def number_tracks(day, city):
    # Almacena las filas del DataFrame donde el valor en la columna 'day' es igual al parámetro day=
    df_filtered = df[df["day"] == day]
    # Filtra las filas donde el valor en la columna 'city' es igual al parámetro city=
    df_filtered = df_filtered[df_filtered["city"] == city]
    # Extrae la columna 'user_id' de la tabla filtrada y aplica el método count()
    track_count = df_filtered["user_id"].count()
    # Devuelve el número de valores de la columna 'user_id'
    return track_count

Etapa 3.6. Llama a `number_tracks()` cuatro veces: una por ciudad en cada uno de los dos días.

In [46]:
# El número de canciones reproducidas en Springfield el lunes
print(number_tracks("Monday", "Springfield"))

15740


In [47]:
# El número de canciones reproducidas en Shelbyville el lunes
print(number_tracks("Monday", "Shelbyville"))

5614


In [48]:
# El número de canciones reproducidas en Springfield el viernes
print(number_tracks("Friday", "Springfield"))

15945


In [49]:
# El número de canciones reproducidas en Shelbyville el viernes
print(number_tracks("Friday", "Shelbyville"))

5895


# Conclusiones <a id='end'></a>

## Escribe tus conclusiones finales sobre el análisis

Redacta un resumen breve y claro de los hallazgos obtenidos durante el proceso de análisis. Tu conclusión debe:

* Mencionar los principales patrones que observaste en los datos.
* Identificar cualquier problema encontrado y cómo lo solucionaste.
* Explicar cómo estas acciones ayudaron a mejorar la calidad del análisis.

Reflexiona también sobre la pregunta central:

> *¿Los datos muestran que el comportamiento de los usuarios —en cuanto a la música que escuchan— varía según la ciudad y el día de la semana?*

Apoya tu respuesta con ejemplos concretos de los resultados obtenidos.


Detalla aquí tus conclusiones.

Encontramos que el numero total de canciones reproducidas en Springfield con 42741 es mucho mayor que en Shelbyville con 18512, por lo tanto Springfield presenta una mayor actividad musical, practicamente mas del doble, siendo los viernes el dia que se eleva un poco el resultado de musica escuchada en Springfield con 15495 reproducciones. se detectaron errores en valores dentro de las columnas "track" "artist" y "genre", los cuales se reemplazaron con "unknown" para evitar errores de conteo, dentro de los generos se reemplazaron generos (en este caso hip, hop y hip-hop) con hiphop para evitar repeticiones innesesarias. Las acciones tomadas nos ayudaron a mejorar el analisis evitando repeticion, conteos fantasma, o trabajar con datos incompletos o inconsistentes.

Si, los datos del comportamiento de usuario son variados segun la ciudad y el dia, la ciudad es donde los datos tienen mas variacion (springfield con 42741 reproducciones y shelbyville con 18512) pero en cuestion de dia, el comportamiento es similar(en lunes con 15740 y 5614; en viernes con 15945 y 5895 reproducciones para Springfield y Shelbyville respectivamente, sin embargo esto no implica que sea igual, siendo variaciones en menor o mayor medida, siguen siendo variaciones y nos otorga informacion del comportamiento de los usuarios.